Fix: Issue #68, 24 hour task rotation thread

joshvasquezr opened the pull request:
Added a new method to `server.c`: 'task_reset_thread()`...
This method runs in a background thread that checks if the current task has been unsolved for > 24 hours. The thread is started in after initial task generation.

malensek commented:
This does not compile.

joshvasquezr commented:
Lol wow, yup, not my finest moment... should compile now, though.

malensek commented:
Ok, yes it does. Think it may make sense for this check to happen less frequently than every 10 seconds? It seems like quite often for something that only happens once per day. Would a minute make sense?

The old code (that's still also present) was correct with respect to using `difftime`, updating the current timestamp, using a more readable timespan instead of the raw number of seconds, and had a meaningful log message with the new block, so I'm not a fan of losing all that with this change.

We aren't exiting cleanly if the thread can't be created and aren't cleaning up its resources on exit.

joshvasquezr commented:
Ok! I believe all issues you mentioned should be fixed and the code still compiles successfully.

malensek commented on server.c:
```diff
@@ -335,13 +331,17 @@ void shutdown_handler(int signo) {
-	    if (now - task_start_time > 86400) {
+            double diff = difftime(now, task_start_time);
+            	    
+	    if (diff > 86400) {
```
24 * 60 * 60                  // 24 hours

malensek commented on server.c:
```diff
@@ -335,13 +331,17 @@ void shutdown_handler(int signo) {
-		    task_start_time = now;
+		    char ts[32];
+		    strftime(ts, sizeof ts, "%Y-%m-%d %H:%M:%S", localtime(&now));
+	            LOG("[RESET] task_reset_thread(): 24 Hours Elapsed - Generating New Task at %s\n", ts);
```
We can take out task_reset_thread(), it's already printed by the LOG.

malensek commented on server.c:
```diff
@@ -413,8 +413,10 @@ int main(int argc, char *argv[]) {
     if (pthread_create(&reset_thread, NULL, task_reset_thread,NULL) != 0) {
-	    perror("Failed to create task reset thread");
-	    exit(1);
+        perror("Failed to create task reset thread");
```
since perror will explain the error, we can do:
perror("pthread_create reset_thread")

malensek commented on server.c:
```diff
@@ -413,8 +413,10 @@ int main(int argc, char *argv[]) {
-	    perror("Failed to create task reset thread");
-	    exit(1);
+        perror("Failed to create task reset thread");
+	task_destroy(); 
+	pthread_mutex_destroy(&lock);
+	return 1;
```
We have a better way of doing cleanup now with our cleanup block, let's use that. May need to check if this initialization needs to be moved later to avoid problems with the cleanup code... but I don't think so?

malensek reviewed (commented):
Looks great. A few style things to tweak and the only small issue is we should reuse the existing cleanup code but otherwise this looks ready to go.

joshvasquezr commented:
Ok! I added all tweaks you suggested and just to be safe, I think there *might* be a possibility that cleaning up before socket_fd is init could cause some weird behavior so I just initialized socked_fd to -1 to avoid anything weird. I also added a boolean flag so that clean up knows whether reset task thread was successfully created...

Finally, there was a merge conflict with checking the 24 hour task reset wtihin client_thread(), so i just deleted that block in the function since task_reset_thread() should handle that successfully...
